### PR TITLE
StokesFO Kokkos Refactor and Boundary Layout Update

### PR DESCRIPTION
**Pull request by mcarlson801**
This PR contains updates to the boundary conditions of the StokesFO problem. All sideset evaluators used in StokesFO have been updated to use the new collapsed sidesets layout while still allowing other problems to use these evaluators with the old sideset layouts. Additionally, the remaining performance intensive evaluators used in StokesFO have been ported to Kokkos. A summary of the changes in this PR is as follows:

- StokesFO sideset evaluators have been updated to allow new or old sideset layouts
- StokesFO evaluators have been fully ported to Kokkos and the total fill time for GPU builds has been reduced significantly
- Added a 'localDOFview' structure to PHAL_Workset that is initialized in STKDiscretization.cpp::computeSideSets(). This structure is a kokkos view that is used in GatherVerticallyContractedSolution to index into the underlying Thyra vector in a coalesced fashion on GPU.
- Added control flow to StokesFOBase so that other problems that extend StokesFOBase use the old sideset layout until they can be updated.
- Views have been added to DiscretizationUtils so that computeBasisFunctionsSide has access to the appropriate data when running on GPU.

### CODE REVIEW

File: src/evaluators/response/PHAL_ResponseSquaredL2DifferenceSide_Def.hpp
Comment (jewatkins) [inline src/evaluators/response/PHAL_ResponseSquaredL2DifferenceSide_Def.hpp:125]: "diff_2"

File: src/evaluators/response/PHAL_ResponseSquaredL2DifferenceSide_Def.hpp
Comment (jewatkins) [inline src/evaluators/response/PHAL_ResponseSquaredL2DifferenceSide_Def.hpp:170]: Need to discuss creating kernels for these

File: src/evaluators/utility/PHAL_ComputeBasisFunctionsSide_Def.hpp
Comment (jewatkins) [inline src/evaluators/utility/PHAL_ComputeBasisFunctionsSide_Def.hpp:436]: I forgot, what was the reason for not making this a kernel?

File: src/evaluators/utility/PHAL_FieldFrobeniusNorm_Def.hpp
Comment (jewatkins) [inline src/evaluators/utility/PHAL_FieldFrobeniusNorm_Def.hpp:257]: need to discuss if kernels are needed for these too

File: src/evaluators/utility/PHAL_MapToPhysicalFrameSide_Def.hpp
Comment (jewatkins) [inline src/evaluators/utility/PHAL_MapToPhysicalFrameSide_Def.hpp:90]: I think cell is not needed in this one

File: src/LandIce/evaluators/LandIce_GatherVerticallyContractedSolution_Def.hpp
Comment (mcarlson801) [reply src/LandIce/evaluators/LandIce_GatherVerticallyContractedSolution_Def.hpp:161]: This was my plan originally but I don't know how I can get access to the LayeredMeshNumbering in the constructor or postRegSetup. I'll look into it to see if there is a way I can get it during discretization.

File: src/LandIce/problems/LandIce_StokesFOBase.cpp
Comment (mcarlson801) [reply src/LandIce/problems/LandIce_StokesFOBase.cpp:171]: Just for debugging. Thanks for catching that. It did confirm that sideMeshSpecs.worksetSize is not enough to determine the actual worksetSize.

File: src/disc/stk/Albany_STKDiscretization.cpp
Comment (mcarlson801) [reply src/disc/stk/Albany_STKDiscretization.cpp:2042]: I do actually check if layered_mesh_numbering exists but a few lines later for some reason. I'll move that up to avoid some unnecessary cycles and see if there is a way to check if it is needed by any evaluators.

Comment (jewatkins) [reply src/disc/stk/Albany_STKDiscretization.cpp:2042]: The latter might be a bit difficult. I don't see a clear parameter that would turn this on so it's probably okay to only construct this if the layered mesh numbering exists. It's probably not a lot of memory compared to the other memory issues we have.

Comment (bartgol) [reply src/disc/stk/Albany_STKDiscretization.cpp:2042]: I don't think the disc has any way to know if the LMN is indeed needed.

File: src/evaluators/utility/PHAL_ComputeBasisFunctionsSide_Def.hpp
Comment (mcarlson801) [reply src/evaluators/utility/PHAL_ComputeBasisFunctionsSide_Def.hpp:436]: I think there was a reason at one point but the way I've rewritten it, it would now be easy to turn it into a kernel. I'll also add one for populating physPointsCell.

Comment (jewatkins) [reply src/evaluators/utility/PHAL_ComputeBasisFunctionsSide_Def.hpp:436]: okay great thanks

File: src/LandIce/evaluators/LandIce_BasalFrictionCoefficientGradient_Def.hpp
Comment (mcarlson801) [reply src/LandIce/evaluators/LandIce_BasalFrictionCoefficientGradient_Def.hpp:243]: BasalFrictionCoefficient and BasalFrictionCoefficientGradient now have Kokkos kernels. The response kernels that you mentioned in other comments here are easy enough to add kernels for so I'm going to add those as well when I have some time tomorrow. Once all kernels are in, I'll do a performance comparison and post the results.

File: src/Albany_StateManager.cpp
Comment (bartgol) [inline src/Albany_StateManager.cpp:579]: Why not erroring out? It seemed appropriate, if the input was unexpected.

File: src/LandIce/evaluators/LandIce_EffectivePressure_Def.hpp
Comment (bartgol) [inline src/LandIce/evaluators/LandIce_EffectivePressure_Def.hpp:100]: This is fine, but for the record, given the small size, I believe keeping both cases in a single kernel (with an if to determine which for loop to execute) would have been good enough. The branch predictor will likely hit the correct branch.

File: src/LandIce/evaluators/LandIce_GatherVerticallyContractedSolution_Def.hpp
Comment (bartgol) [reply src/LandIce/evaluators/LandIce_GatherVerticallyContractedSolution_Def.hpp:161]: I think the struct LayeredMeshNumbering is probably putting a lot of info in the same struct. Some, belongs to the MeshStruct itself (which is where the LayeredMeshNumbering ptr is currently stored), but the number of layers is more of a mesh specs kind of info (much like the workset size).

Perhaps we could add a couple of vars to the MeshSpecs:
```
bool is_layered;
int num_layers;
```
Or even just `num_layers`, with the convention that, say, a negative number means not-layered. The number of layers can also be retained in the LayeredMeshNumbering struct, for ease of use (only one struct to lookup layered mesh info). But having this one bit of info in the mesh specs might help to do this kind of setup in the ctor phase.

File: src/disc/stk/Albany_STKDiscretization.cpp
Comment (bartgol) [inline src/disc/stk/Albany_STKDiscretization.cpp:2041]: Do you set '6' cause you're thinking about quadratic triangles? I don't think we support quadratic elements in Albany. But then, if we did, 6 would not fit quadratic Quads, which would show up when using prisms.

You could be more tight, and loop over the cell_topo sides, and count their nodes. That will tell you exactly what 'max_side_nodes' should be. I suspect 6 is an overshoot for most of our cases. A tighter upper bound might help with caching.

File: src/evaluators/interpolation/PHAL_DOFCellToSide.hpp
Comment (bartgol) [inline src/evaluators/interpolation/PHAL_DOFCellToSide.hpp:62]: So you are only producing an output with collapsed layout for the "P0-scalar" and P1-scalar/vector cases, is that correct?

That's fine, for now, but it's worth making a TODO action item, to propagate collapsed layout also to P0 vector/tensor and P1-tensor quantities.

Note: this can be done when the non-collapsed side layouts are completely removed.

File: src/problems/Albany_Layouts.cpp
Comment (bartgol) [inline src/problems/Albany_Layouts.cpp:207]: I was just thinking, that maybe we could switch to `PHX::Layout` instead of `PHX::MDALayout`. The former allows to defer the specification of the extents to a later time. For us, that'd be after the whole mesh has been created, so that we know the exact number of sides on the sideset, and can allocate smaller side layouts.

The main drawback is that with dynamic layouts, you _need_ to call `setFieldData` for each required/evaluated field in `postRegistrationSetup`, while now fields are allocated at construction time, so the `setFieldData` step is superfluous (and sometimes I don't even do it in the evaluators I write). Also, you would not be able to get the number of nodes or quad points from a layout at construction time inside the evaluators (like we sometimes do).

Just a thought, for later.

P.s: obviously, `PHX::Layout` still inherits from `PHX::DataLayout`, so the modifications would be minimal.

**Review by bartgol — APPROVED** [review]
Monstruous PR! Great work. I had some comments, but I believe they are just thoughts, for future development and/or to spark discussions.

File: src/Albany_StateManager.cpp
Comment (mcarlson801) [reply src/Albany_StateManager.cpp:579]: The exception that I removed here was copy and pasted from the non-collapsed implementation and only throws if there is a layout with only one field. Since the old sideset layouts have to have two fields (Cell, Side) at minimum, there was no way it could have only one field. However, with the collapsed layout, it is fine to have a single field (Side). From what I could gather, for a single field layout, all we need to do here is get the dimensions.

File: src/disc/stk/Albany_STKDiscretization.cpp
Comment (mcarlson801) [reply src/disc/stk/Albany_STKDiscretization.cpp:2041]: Yeah, this was a bandaid thing that I didn't get around to fixing. I'll loop over the sides to get the tightest bound here.

File: src/Albany_StateManager.cpp
Comment (bartgol) [reply src/Albany_StateManager.cpp:579]: I'm just weary of 'else' statements, cause they catch whatever didn't fit in the previous if(s). Maybe we can put a test for exception, ensuring that the rank is indeed 1? Right now, you didn't add, say, `qp_tensor3` as a collapsed layout to Albany::Layout. But if tomorrow someone did it, and used it, adding a check in here might save hours of debugging.

Comment (bartgol) [reply src/Albany_StateManager.cpp:579]: Note: alternatively, you could add a check on the layout rank at the beginning of the fcn, and support only rank<=4 for now.

Comment (mcarlson801) [reply src/Albany_StateManager.cpp:579]: That's a good point. I'll rewrite this to make sure nothing gets through the cracks. I assume that means an else-if for rank 1 and then throw an exception in the else but I'll double check to make sure that covers all cases.

**Comment by mperego** [comment]
@mcarlson801 @jewatkins  I'm going to start working on some evaluators modified by this PR. Do you suggest that I base my work on this branch?

**Comment by jewatkins** [comment]
> @mcarlson801 @jewatkins I'm going to start working on some evaluators modified by this PR. Do you suggest that I base my work on this branch?

If they're sideset evaluators, I would say yes. Although if you need things in master soon maybe not. I think there's still a few things Max needs to fix and it would be good to have the GPU performance tests up before merging this. See https://github.com/SNLComputation/Albany/issues/661 and https://github.com/trilinos/Trilinos/issues/8727.

**Comment by mperego** [comment]
> > @mcarlson801 @jewatkins I'm going to start working on some evaluators modified by this PR. Do you suggest that I base my work on this branch?
> 
> If they're sideset evaluators, I would say yes. Although if you need things in master soon maybe not. I think there's still a few things Max needs to fix and it would be good to have the GPU performance tests up before merging this. See #661 and [trilinos/Trilinos#8727](https://github.com/trilinos/Trilinos/issues/8727).

I see. Yes, I'm going to work on the basal friction evaluator.

**Comment by mperego** [comment]
@mcarlson801 when I run the testsuit using this branch, a lot of tests fail with this error:
```
3: p=3: *** Caught standard std::exception of type 'std::logic_error' :
3: 
3:  /ascldap/users/mperego/Workspace/albany/sources/albany-src/src/disc/stk/Albany_STKDiscretization.hpp:210:
3:  
3:  Throw number = 1
3:  
3:  Throw test that evaluated to true: true
3:  
3:  Local DOF view was requested but it has not been created.
```
Have you seen that?

**Comment by mcarlson801** [comment]
> @mcarlson801 when I run the testsuit using this branch, a lot of tests fail with this error:
> 
> ```
> 3: p=3: *** Caught standard std::exception of type 'std::logic_error' :
> 3: 
> 3:  /ascldap/users/mperego/Workspace/albany/sources/albany-src/src/disc/stk/Albany_STKDiscretization.hpp:210:
> 3:  
> 3:  Throw number = 1
> 3:  
> 3:  Throw test that evaluated to true: true
> 3:  
> 3:  Local DOF view was requested but it has not been created.
> ```
> 
> Have you seen that?

@mperego Hmm, I must not have run unit tests since adding that exception. I'll look into this and give you an update shortly.

**Comment by mcarlson801** [comment]
> @mcarlson801 when I run the testsuit using this branch, a lot of tests fail with this error:
> 
> ```
> 3: p=3: *** Caught standard std::exception of type 'std::logic_error' :
> 3: 
> 3:  /ascldap/users/mperego/Workspace/albany/sources/albany-src/src/disc/stk/Albany_STKDiscretization.hpp:210:
> 3:  
> 3:  Throw number = 1
> 3:  
> 3:  Throw test that evaluated to true: true
> 3:  
> 3:  Local DOF view was requested but it has not been created.
> ```
> 
> Have you seen that?

@mperego I figured out what the issue is. I added that exception to "getLocalDOFViews" so that we knew if a workset was grabbing that view if it didn't need it. However, in Albany_Application, "getLocalDOFViews" is always called for every workset regardless if it is needed.

I've reverted this change temporarily so that I'm not interrupting your workflow but a proper fix is coming later today in addition to a fix for an issue that only shows up on GPU builds.

**Comment by mcarlson801** [comment]
> @mcarlson801 when I run the testsuit using this branch, a lot of tests fail with this error:
> 
> ```
> 3: p=3: *** Caught standard std::exception of type 'std::logic_error' :
> 3: 
> 3:  /ascldap/users/mperego/Workspace/albany/sources/albany-src/src/disc/stk/Albany_STKDiscretization.hpp:210:
> 3:  
> 3:  Throw number = 1
> 3:  
> 3:  Throw test that evaluated to true: true
> 3:  
> 3:  Local DOF view was requested but it has not been created.
> ```
> 
> Have you seen that?

@mperego It looks like this change wasn't enough to pass unit tests. I'm working on a more comprehensive fix and I'll let you know as soon as this branch is testing clean. Sorry for the inconvenience, I'll try to get this working ASAP.

**Comment by mperego** [comment]
@mcarlson801  Thanks. Do you mind if I cherry-pick your last commit on the my version where I just merged the master and force-push? Otherwise I need to do the merging again

**Comment by mcarlson801** [comment]
> @mcarlson801 Thanks. Do you mind if I cherry-pick your last commit onto my version where I just merged the master and force-push? Otherwise I need to do the merging again

@mperego Yeah, if that works best for you, go for it. I'll work around whatever you decide to do, shouldn't be a problem.

**Comment by mperego** [comment]
@mcarlson801 It's probably best if I hold off while you are fixing things. I won't mess up with this branch

File: src/evaluators/utility/PHAL_FieldFrobeniusNorm_Def.hpp
Comment (mperego) [inline src/evaluators/utility/PHAL_FieldFrobeniusNorm_Def.hpp:96]: @mcarlson801  I think this should be changed to `dl->node_vector_sideset` when `useCollapsedSideset` is `true`. Same in the next two if statemets

Comment (mcarlson801) [reply src/evaluators/utility/PHAL_FieldFrobeniusNorm_Def.hpp:96]: Yeah, this does seem to be a mistake on my part. The fix will come in with my next commit.

File: src/LandIce/evaluators/LandIce_BasalFrictionCoefficient_Def.hpp
Comment (mperego) [inline src/LandIce/evaluators/LandIce_BasalFrictionCoefficient_Def.hpp:315]: @mcarlson801  I think you need to swap the Tag types of this and the following function.

File: src/evaluators/state/PHAL_SaveSideSetStateField_Def.hpp
Comment (mperego) [inline src/evaluators/state/PHAL_SaveSideSetStateField_Def.hpp:66]: @mcarlson801  The two TEUCHOS_TEST_FOR_EXCEPTION below, need to be modified for collapsed side sets (`.size() < 2` and `.name(1) `)

File: src/evaluators/state/PHAL_SaveSideSetStateField_Def.hpp
Comment (mperego) [inline src/evaluators/state/PHAL_SaveSideSetStateField_Def.hpp:451]: @mcarlson801 this `switch` needs to be modified for collapsed side sets (`dims.size()` is either 2 or 3), also when you access `dims`, you need to reduce the index by one.

File: src/evaluators/state/PHAL_SaveSideSetStateField_Def.hpp
Comment (mperego) [inline src/evaluators/state/PHAL_SaveSideSetStateField_Def.hpp:401]: @mcarlson801 this `switch` needs to be modified for collapsed side sets (`dims.size()` is either 2 or 3), also when you access `dims`, you need to reduce the index by one.

**Comment by mperego** [comment]
@mcarlson801 I notice that the `StokesFOThermoCoupled` problem is still using the non collapsed sidesets. Are you planning to change that in this PR or is it separated by this? 
Sorry for keeping pinging you. As I mentioned I'm working at different evaluators that are based on sideset and I'd rather do that only once using the new layout and Kokkos implementation.

**Comment by mcarlson801** [comment]
> @mcarlson801 I notice that the `StokesFOThermoCoupled` problem is still using the non collapsed sidesets. Are you planning to change that in this PR or is it separated by this?
> Sorry for keeping pinging you. As I mentioned I'm working at different evaluators that are based on sideset and I'd rather do that only once using the new layout and Kokkos implementation.

@mperego The only problem that will be updated in this PR is StokesFO. The other problems that inherit from StokesFOBase will be changed in a later PR.

Once I have tested this next (and hopefully final) commit for this PR, I will take a look at the StokesFOThermoCoupled problem to see what is needed to get it working with the collapsed sidesets. It may be that all of the evaluators for StokesFOThermoCoupled are ready for the new layout and in that case I could just switch a flag and it would use the new layouts. I'll update you when I know more.

**Comment by mperego** [comment]
OK, thanks!  I'll start working on that. For sure `w_Resid` evaluator needs to be converted to have StokesFOThermo coupled to work.

**Comment by mperego** [comment]
@mcarlson801 B.t.w. I confirm that  `w_Resid` is the only evaluator that needs to be changed in order to use the collapsed layout in `StokesFOThermoCopled`.

**Comment by mperego** [comment]
@mcarlson801 thanks for all this work. I've merged master into the branch. @mcarlson801 @jewatkins is anything preventing this from being merged? More testing?

**Comment by jewatkins** [comment]
> @mcarlson801 thanks for all this work. I've merged master into the branch. @mcarlson801 @jewatkins is anything preventing this from being merged? More testing?

I think there's still some lingering issues that Max was going to address today.

**Comment by mcarlson801** [comment]
> > @mcarlson801 thanks for all this work. I've merged master into the branch. @mcarlson801 @jewatkins is anything preventing this from being merged? More testing?
> 
> I think there's still some lingering issues that Max was going to address today.

The PR now has all the changes I was planning to make and is currently passing all unit tests, the only thing left is testing for GPU builds and then I think Jerry wants to look over it once more for a final review.

**Comment by jewatkins** [comment]
I think the trilinos gpu fix is in develop so I'll go ahead and do a final review/test and push if everything looks good.

File: src/evaluators/utility/PHAL_ComputeBasisFunctionsSide_Def.hpp
Comment (jewatkins) [inline src/evaluators/utility/PHAL_ComputeBasisFunctionsSide_Def.hpp:246]: this is a race condition if multiple cells touch a single "sideset index on side". I'm not sure if that's the case here.

**Review by jewatkins — CHANGES_REQUESTED** [review]
A few more things. I think we're close. All performance tests pass except one (I think it's a known stability issue which I need to look into).

**Comment by mperego** [comment]
@mcarlson801 what's the ETC for this PR? We have enother PR that is ready to be merged in. My understanding is that they are complitely orthogonal. Should we merge the other PR first?

**Comment by jewatkins** [comment]
> @mcarlson801 what's the ETC for this PR? We have enother PR that is ready to be merged in. My understanding is that they are complitely orthogonal. Should we merge the other PR first?

There's a few minor edits and Max is seeing nondeterministic behavior in a performance test that needs to be tracked down. I don't see an issue with merging the other PR first.

File: src/LandIce/evaluators/LandIce_BasalFrictionCoefficient_Def.hpp
Comment (bartgol) [reply src/LandIce/evaluators/LandIce_BasalFrictionCoefficient_Def.hpp:252]: I'm not sure. The scalar types of the two fields are different, so deep_copy might either not compile, or do a memcpy (which would be even worse). I guess we could try and though.

File: src/LandIce/evaluators/LandIce_GatherVerticallyContractedSolution_Def.hpp
Comment (mcarlson801) [reply src/LandIce/evaluators/LandIce_GatherVerticallyContractedSolution_Def.hpp:175]: Since this change is okay for now, I'm going to resolve this one.

**Comment by mperego** [comment]
> > @mcarlson801 what's the ETC for this PR? We have enother PR that is ready to be merged in. My understanding is that they are complitely orthogonal. Should we merge the other PR first?
> 
> There's a few minor edits and Max is seeing nondeterministic behavior in a performance test that needs to be tracked down. I don't see an issue with merging the other PR first.

@jewatkins OK! I merged the other PR. Sorry for the nondeterministic behavior.. those are painful to fix.

File: src/LandIce/evaluators/LandIce_BasalFrictionCoefficient_Def.hpp
Comment (jewatkins) [reply src/LandIce/evaluators/LandIce_BasalFrictionCoefficient_Def.hpp:252]: Good point, I did not see the different scalar type. I'm not sure if sacado has the overloads for this case. I would hope there's something in place to prevent a direct memcpy but maybe there's no such guard.

File: src/evaluators/interpolation/PHAL_DOFCellToSide.hpp
Comment (mcarlson801) [reply src/evaluators/interpolation/PHAL_DOFCellToSide.hpp:62]: I'll add a TODO here just to be safe but yes, you have that right and my plan is to address this during the removal of the non-collapsed layouts.

File: src/evaluators/utility/PHAL_ComputeBasisFunctionsSide_Def.hpp
Comment (mcarlson801) [reply src/evaluators/utility/PHAL_ComputeBasisFunctionsSide_Def.hpp:246]: The sideset indices for each side are guaranteed to be disjoint sets so this one is safe.

Comment (mperego) [reply src/evaluators/utility/PHAL_ComputeBasisFunctionsSide_Def.hpp:246]: Maybe it's worth adding a comment stating that.

File: src/Albany_StateManager.cpp
Comment (mcarlson801) [reply src/Albany_StateManager.cpp:580]: I revisited this section of the code to see what needed to be done here. The else here is covering the case where stateRef.entity is not nodal. There isn't really anything in this if-else that depends on layout size so I'm not sure it actually makes sense to do an exception here based on layout rank beyond checking for the bad case rank == 1 which is no longer relevant for collapsed layouts. 

It seems like to me that there are already exceptions earlier in the function to make sure we aren't getting unexpected input so as far as I can tell nothing needs to be added here. Let me know if I am missing something here since I am not super familiar with the stateRef entity stuff.

File: src/LandIce/evaluators/LandIce_BasalFrictionCoefficient_Def.hpp
Comment (mcarlson801) [reply src/LandIce/evaluators/LandIce_BasalFrictionCoefficient_Def.hpp:252]: It looks like we will have to mess with this a bit more to get it to compile so I'm marking resolved for now and we can revisit in the future.

**Comment by mperego** [comment]
@mcarlson801 thanks!!

**Pull request merged**
